### ISSUE TITLE
API for manipulating token-level input embeddings

### ISSUE THREAD

**Issue by ringohoffman** [ggerganov/llama.cpp#4537]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Feature Description

API to allow manipulation of token-level input embeddings.

# Motivation

I work for [Protopia AI](https://protopia.ai/llms-ai/), a company that offers an LLM privacy solution that works by transforming LLM input embeddings. We have a client that uses llama.cpp, and we are interested to see how our PyTorch based solution can integrate with llama.cpp. I have been looking at [llama-cpp-python](https://github.com/abetlen/llama-cpp-python) as an avenue to understand llama.cpp's APIs.

I've tried to understand the current embedding API with little luck:

[llama-cpp-python discussion: Embedding for sentence](https://github.com/abetlen/llama-cpp-python/discussions/503)
[llama-cpp-python discussion: Load 70b model only once -- for embedding and for completion](https://github.com/abetlen/llama-cpp-python/discussions/592)

[llama.cpp discussion: What exactly does llama_get_embeddings return?](https://github.com/ggerganov/llama.cpp/discussions/3643)
[llama.cpp issue: Bug: Invalid Embeddings if GPU offloaded (CUDA)](https://github.com/ggerganov/llama.cpp/issues/3625)
[llama.cpp discussion: Where does the embedding come from?](https://github.com/ggerganov/llama.cpp/discussions/3243)
[llama.cpp discussion: How to use embedding ?](https://github.com/ggerganov/llama.cpp/discussions/2180)
[llama.cpp discussion: How do I get input embeddings?](https://github.com/ggerganov/llama.cpp/discussions/1013)
[llama.cpp discussion: How to get and modify weights between activated neurons?](https://github.com/ggerganov/llama.cpp/discussions/2654#discussioncomment-6763602)

How much work would it be to expose access to the token-level input embeddings for manipulation at inference? Is this possible using llama.cpp's existing APIs?

**Comment by abb128** [comment]
This was very confusing to me initially, but I found the following (ugly) method to work

You can extract the full input embed matrix like so
```c
    std::vector<float> embeddings;
    embeddings.resize(llama_n_embd(model) * llama_n_vocab(model));

    auto tensor = llama_get_model_tensor(model, "token_embd.weight");
    ASSERT(tensor);

    if(tensor->type != GGML_TYPE_F32) {
        ggml_internal_get_type_traits(tensor->type).to_float(tensor->data,
                                                             embeddings.data(),
                                                             embeddings.size());
    } else {
        ASSERT((tensor->ne[0] * tensor->ne[1]) == embeddings.size());
        memcpy(embeddings.data(), tensor->data, embeddings.size() * sizeof(float));
    }
```

You can then convert token id to embeds like so
```c
                    const float *embeds = embeddings.data() +
                                 (t.token * n_embd);
```

Then use the embeds field of llama_batch to input this embedding or a modified embedding based on this, in my experience I think you can only decode one token at a time, this may have changed since I last updated
